src/main/java/org/broadinstitute/hellbender/utils/hmm/ForwardBackwardAlgorithm.java

```diff
@@ -0,0 +1,432 @@
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Implementation of the interface {@link Result} returned by the {@link #apply} method.
+     * @param <D> the observed data type.
+     * @param <T> the observation time/position type.
+     * @param <S> the hidden state type.
+     * @return never {@code null}.
+     */
+    private static class ArrayResult<D, T, S> implements Result<D, T, S> {
+
+        private final List<D> data;
+
+        private final List<T> positions;
+
+        private final HiddenMarkovModel<D, T, S> model;
+
+        private final Object2IntMap<T> positionIndex;
+        private final Object2IntMap<S> stateIndex;
+
+        private final int width;
+
+        private final double[] logForwardProbabilities;
+
+        private final double[] logBackwardProbabilities;
+        private final double logDataLikelihood;
+
+        private ArrayResult(final List<D> data, final List<T> positions,
+                            final HiddenMarkovModel<D, T, S> model,
+                            final double[] logForwardProbabilities,
+                            final double[] logBackwardProbabilities) {
+            this.data = Collections.unmodifiableList(new ArrayList<>(data));
+            this.positions = Collections.unmodifiableList(new ArrayList<>(positions));
+            this.model = model;
+            this.positionIndex = composeIndexMap(this.positions);
+            this.stateIndex = composeIndexMap(model.hiddenStates());
+            this.width = stateIndex.size();
+            this.logBackwardProbabilities = logBackwardProbabilities;
+            this.logForwardProbabilities = logForwardProbabilities;
+            this.logDataLikelihood = calculateLogDataLikelihood(logForwardProbabilities, logBackwardProbabilities
+                , this.stateIndex.size());
+        }
+
+        /**
+         * Calculates the data likelihood in log scale.
+         * <p>
+         *     This value can be obtain by adding up the posterior probabilities at any position; they are all supposed to
+         *     be the same.
+         * </p>
+         * <p>
+         *     This implementation uses always the first position.
+         * </p>
+         * @param logForwardProbabilities the log forward probabilities array.
+         * @param logBackwardProbabilities the log backward probabilities array.
+         * @param width number of hidden states.
+         * @return a valid probability in log scale (between -Inf and 0 inclusive).
+         */
+        private static double calculateLogDataLikelihood(final double[] logForwardProbabilities,
+                                                         final double[] logBackwardProbabilities,
+                                                         final int width) {
+            if (logForwardProbabilities.length == 0) {
+                return 0;
+            } else {
+                final double[] values = new double[width];
```

Comment:
If probabilities are 2-D arrays this code becomes `return logSumLog(arraySum(forward[0], backward[0]))`.